### DISCUSSION TITLE
proper motion in which module added to stars position?

### DISCUSSION THREAD

**Discussion by vreijs** [Stellarium/stellarium#3101] [Q&A]
Can someone tell me in which module the proper motion of a star is being included in de declination/right ascension?
I am not that used to github, so searching in the source-code has not been successful for me.
Thanks for the help.

Why I am asking is: Is the RA's proper motion from the star database changed with a factor 
1/cos(declination)? https://www.fxsolver.com/browse/formulas/Magnitude+of+proper+motion+%28%CE%BC%CE%B1%2A%29

I hope to find that in the source code.

Thanks for your help.

All the best,

Victor

**Comment by gzotti** [comment]
Hi! The star catalog (created around 2006) is described in Appendix C of the User Guide. The relevant classes are Star* and ZoneArray. I am not aware of more documentation than those files. Analysis and reverse documentation more than welcome!

**Comment by vreijs** [comment]
Thanks Georg,
I found the proper motion gathering at StarMgr::populateHipparcosLists() in; https://github.com/Stellarium/stellarium/blob/master/src/core/modules/StarMgr.cpp
	712		PMData properMotion = getProperMotion(s->getHip());
	713		float pmX = properMotion.first;
	714		float pmY = properMotion.second;
	715		float pm = 0.001f * std::sqrt((pmX*pmX) + (pmY*pmY));

Somewhere else in the code pmY/cos(declination) (conversion of spherical coordinates) needs to be done as the actual star declination is not yet known when generating this StarMgr::populateHipparcosLists()? IMHO also pm (and its angle) is depending on this declination (pm = 0.001f * std::sqrt((pmX*pmX) + (pmY/cos(declination)*pmY/cos(declination)));
If someone knows where the star position is calculation in the source code, can you please provide a link to that code module? 
I want to look at this code and see if I can find the reason why eg Rigel Kentaurus is so much off in coordinates (it has large proper motion).

I am not accustomed to how such a github code environment works (or how one can search easily in the files). Sorry.

All the best,

Victor

**Reply by Atque** [reply]
>I want to look at this code and see if I can find the reason why eg Rigel Kentaurus is so much off in coordinates (it has large proper motion).

Please take a look at Cassiopeia too, if you find anything. It is very corrupted compared to other open-source programs such as Cartes du Ciel.

![cdc_vs_stellarium](https://user-images.githubusercontent.com/37028368/225092276-2051ac6d-29bc-4e76-9b43-80ac6e537b35.png)

**Reply by vreijs** [reply]
I will also look at it. I hope that it is a simple change that might also solve your star position (Cassiopheia also has a high abs(declination), like Rigel). Although the proper motion of Cassiopheia stars is not so high as Rigel K. so I don't know if it will solve it for you... First need to knwo where the code is that determines the stars position based on proper motion.

**Reply by vreijs** [reply]
Hello Atque,
Was is the date of the picture, say around today? For around today; I don't see much difference in Decl and RA of Stellarium and SwissEphemeris (the one that I use for comparing) for Caph and Ruchdah. Did you check the declinations/RAs of Carte Ciel with Stellarium's (at J2000)?

**Reply by Atque** [reply]
The screenshot above is taken at 10,000 BCE. The values shown in the infostring are hard-coded AFAIK, and are not used for computing actual positional data.

**Reply by vreijs** [reply]
Aha, proper motion big over 12000years becomes big, so it becomes closer Rigel K.'s (over 3000years, over which I am looking). So perhaps my idea might help you. But not sure yet...

**Reply by vreijs** [reply]
By the way, I have the feel that Cart Du Ciel 4.2.1 is correct (as also Rigel K. has in Carte du Ciel the same coordinates as in Swiss Ephemeris and SkyMap Pro 12).

**Reply by Atque** [reply]
Cartes du Ciel uses the most correct data available (even with three-dimensional correction), although the individual components of Rigil Kentaurus will drift apart in the far future or past, due to lack of double star modeling.

**Reply by vreijs** [reply]
So now I need to find the reason why it is different;-) I hope to work on that in the coming weeks, if someone provides me help to find the right code module of the calculation of stellar position.

**Reply by vreijs** [reply]
Here are some comparison for Rigel K. Looking at swephR&SwissEph (using the same core-code), SkyMap Pro 12, Cartes du Ciel 4.2.1 and Stellarium:
![image](https://user-images.githubusercontent.com/17218070/225162741-44fc4ceb-ee68-496e-a4b8-3b248459b98a.png)

**Reply by vreijs** [reply]
In Stellarium: star with small proper motion [smaller than Rigel K.) don't differ much from results of the other celestial programs

**Comment by alex-w** [comment]
> I found the proper motion gathering at StarMgr::populateHipparcosLists() in;

Wrong place. This is method for informer.

**Reply by vreijs** [reply]
Agreed in my original question. But what is the right place? Can you help me?

**Reply by alex-w** [reply]
Please see `ZoneArray.cpp` file, but I should add very important notes: 
* our catalog has only 2 componets for proper motions
* our catalog has inner representation the values of proper motion and coordinates for stars (see `util/Parse*.C` and `util/MakeCombinedCatalogue*.C` files - this is generator `*.cat` files)

**Comment by vreijs** [comment]
Thanks Alexander and also Georg, 

For me 2 components for proper motion (archaeoastonomy upto say 5000BCE) is good enough.

I now (sorry Georg) see in ZoneArray::initTriangle
something that might related to proper motion (as in literature it is also sometimes referred to mu:  μ 2 = μ δ 2 + μ a* 2 ), eg:
57    mu0 = (c0-z.center)*z.axis0;
58    mu1 = (c0-z.center)*z.axis1;
59    f = 1.f/std::sqrt(1.f-mu0*mu0-mu1*mu1);
60   h = std::fabs(mu0)*f;

I still don't see a reference to the correction needed for spherical coordinates (to get μ a*: aka u a/cos(declination).

Is this correction done somewhere else? Sorry to ask these basic module questions. 

All the best,

Victor

**Reply by vreijs** [reply]
This the code in opensource software SwissEphemeris (sweph.c) :
  ra_pm *= DEGTORAD;
  de_pm *= DEGTORAD;
  ra_pm /= cos(de); /* catalogues give proper motion in RA as great circle */

**Reply by alex-w** [reply]
```
void getJ2000Pos(const ZoneData *z,float movementFactor, Vec3f& pos) const
	{
		pos = z->axis0;
		pos*=(static_cast<float>(getX0())+movementFactor*getDx0());
		pos+=(static_cast<float>(getX1())+movementFactor*getDx1())*z->axis1;
		pos+=z->center;
	}
```

**Reply by vreijs** [reply]
I now downloaded all Stellarium files and look for string (i was first on the Github, but it ha snot search function).
So I see reference of GetJ2000Pos in:
ZoneArray. cpp, StarWrapper.hpp and Star.hpp
I tried to browse thing (and also tried to understand what is done, but different without comments; as Georg already warned in some way).

But I have not been able to find a reference for this cos(star's-declination). I think the movementfactor is the epoch-to-reference-date, and Dx0 and Dx1 are the proper motions. And I think these proper motions are read from the Star object (Star.hpp), and I believe these numbers do not include somewhere this cos(star's-declination). Is my analysis correct?

Thanks for your patience.

All the best,

Victor

**Reply by gzotti** [reply]
I can only guess that you will find anything around how PM is packed in the files in the util subdirectory that Alexander pointed out to you. The labeling as "axis0, axis1" indicates that these are most likely not RA/DE, but have to do with the zone system. But I have never fully reverse-engineered the star catalog myself. A huge problem I have suffered from for years was underdocumentation. Yes, docs need time, but help the future developers and maintainers!
As pointed out in the paper referenced in Appendix F of the Guide, Stellarium's stellar PM has some problem. CdC is recommended as open-source program for star positions in prehistory.

**Reply by vreijs** [reply]
Is it possible to contact the past programmer? I think my first mentioning this cos issue is a few years ago, as I find it a real pity that in principle all star positions are wrong (in the past and the future; except around 200CE), certainly because Stellarium is recommended by some to be a great tool for archaeoastronomy. Correct stars positioning is in some way the simplest in all astronomy. 
Good to hear there is a warning in the Guide, but still.
I am willing to discuss this further (even diving into the software), IF I can talk with someone who was involved in coding it in the past or who wants to be involved in the future.

Any takes, who are reading this message?

**Reply by gzotti** [reply]
Google and a chain of links ultimately brought me to https://sourceforge.net/p/glunatic/. Maybe this helps.

**Comment by vreijs** [comment]
I think I know how it should work (that is why I gave this extra code-line from the opensource Swiss Ephemeris, which produces the correct star positions).
<Swiss Ephemeris [https://www.astro.com/swisseph/swephinfo_e.htm ] does not include a GUI, it is only the kernel to calculate positions of celestial bodies using DE430/431>

I have perhaps another possible idea to see if the proper motion is wrongly implemented and if my idea solves it. 

I understand (perhaps wrongly or correctly) that the textual star information is calculated in another way than the plotting of the star: say two totally different tracks... Both use star catalogue, but a different representation of it.
I think the plotting uses a just-enough-accurate table (derived from the catalogue: for instance resulting in: dx0 en dx1).
I got perhaps the idea that StarMgr code is used for printing the textual star information on screen, for instance: pmX and pmY.

Is these two totally different tracks exist, would it be possible only to change the textual star information only (by using pmY=pmY/cos(star's-declination) or is it pmX=pmX/cos(star's-declination): any way it should be related to RA)? 
This change could at least check if it is about this cos-function (without first understanding how the star plotting goes).

What about this idea, or do I fully misunderstand these two possible tracks.

All the best,

Victor

**Reply by gzotti** [reply]
I cannot dive into this part of the code in the coming weeks and don't want to speculate. AFAICR the printing of PM values was introduced around 2016, and according to release notes was last updated with 0.21.2. To fully understand the details of the star catalog, I would start with analyzing the code for catalog creation which transforms from the source catalogs to the ZoneArray structure.

**Reply by vreijs** [reply]
If we could get a student to work on this important (and for me essential) thing?

**Comment by vreijs** [comment]
It looks that my understanding of the structure of Stellarium is not enough to debug further the (incorrect) stars' position. It looks that for dates before and after J2000, the position is (slightly) off: The bigger the time into the past/future and the larger the proper motion, the larger this error is. See examples earlier in this Q&A thread.
This might be due to a correction ('cos(Star's-declination)') related to spherical coordinates.

I have not been able to find such a proper motion related correction in the code. I found a few locations relating to proper motion:
struct Star.hpp:
		pos = z->axis0;
		pos*=(static_cast<float>(getX0())+movementFactor*getDx0());
		pos+=(static_cast<float>(getX1())+movementFactor*getDx1())*z->axis1;

PopulateHipparcosLists (StarMagr.cpp):
			PMData properMotion = getProperMotion(s->getHip());
			float pmX = properMotion.first;
			float pmY = properMotion.second;
			float pm = 0.001f * std::sqrt((pmX*pmX) + (pmY*pmY));
			if (qAbs(pm)>=pmLimit)

ZoneArray.cpp
	float mu0,mu1,f,h;
	mu0 = (c0-z.center)*z.axis0;
	mu1 = (c0-z.center)*z.axis1;
	f = 1.f/std::sqrt(1.f-mu0*mu0-mu1*mu1);

Printing on screen (StarWrapper.cpp):
		float dx = properMotion.first;
		float dy = properMotion.second;
		float pa = std::atan2(dx, dy)*M_180_PIf;

Nowhere I see this correction, which should include 'cos(star's-declination)'. Of course I could have missed the source file, so it looks I have not enough knowledge about the structure of the source files to find it. 
Of course I am not 100% sure that this missing correction will solve the found issues with stars.

I though get also the feel that this problem is mention in Appendix F of the Guide, on Stellarium's stellar PM. CdC is recommended as open-source program for star positions in prehistory.
I personally think that having stars at the correct position in the past and future is essential for a program like Stellarium. The stars position is also the 'easier' calculable item in astronomy, IMHO.

I would like to change this Q&A post into a bug report. Is that possible?
I am willing to test and think about this bug, but I have not enough experience to program things in this github environment, sorry.

All the best,

Victor

**Reply by gzotti** [reply]
You can click on "Create issue from this discussion". However, we have known about "some problem with the catalog" for years, so it will not speed up any solution. Just that it takes more than a few days to solve. Besides all the other 300 little bugs and issues which likely annoy more people during their nights out there in present times. 

What I can IMAGINE is that during creation of the ZoneArray the coordinates and proper motions are processed into the center, axis0 and axis1 elements for maximum efficiency during rendering. (Just compare screen update rates between CdC and Stellarium.) Whatever this means. If it is as easy to solve all this as fixing a stupid cosine error during catalog creation, somebody could apply this fix to the encoding program in the utils directory and reprocess the star catalog. From my side, this is not going to happen in March or even before summer, sorry. 

7d34689b3010fd3325149ab5419c4ab0b8a7ed73 introduced _display_ of PM values from the original catalog (?), read from a separate file. It does not decode PM values from the catalog, and it does not apply PM to the star positions.
